### ISSUE TITLE
GPU delegate gives different result from CPU

### ISSUE THREAD

**Issue by ShiyongL** [tensorflow/tensorflow#40154]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution:Linux Ubuntu 18.04
- Mobile device if the issue happens on mobile device: Pixel 3 XL
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): r2.2.0 and master
- Python version: 3.6
- Bazel version (if compiling from source): 3.0.0
- GCC/Compiler version (if compiling from source): default
- CUDA/cuDNN version: 10.0
- GPU model and memory: 

**Describe the current behavior**
GPU delegate gives very different result from CPU. 
I was able to hard-code the source (in tensorflow/lite/delegates/gpu/common/model_builder.cc) to allow some operations to be delegated to GPU. Out of 270+ operations: 
- Delegating just one Conv_2d will produce very similar result as the one by CPU only. 
- Delegating a few more operations seem to produce bigger difference. 
- Delegating just the first MUL operation will produce very different result. 

**Describe the expected behavior**
result should be close

**Standalone code to reproduce the issue**
I personally hijacked tflite tools/benchmark code and give an sample image as deterministic input, instead of random input. 
I would love to provide my code change if it helps. 

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The tflite model was converted from InsightFace/ArcFace MXNet model
https://github.com/deepinsight/insightface/wiki/Model-Zoo (3.2 model)
link to download the tflite
https://drive.google.com/file/d/1pJX2I8btskVy-QHiF-mcUF7HF6ZFaQuf/view?usp=sharing

Also attached the graph of above model:
[visualized_official_arcface_no_sub.zip](https://github.com/tensorflow/tensorflow/files/4731201/visualized_official_arcface_no_sub.zip)

**Comment by ShiyongL** [comment]
@srjoglekar246 @impjdi Thanks a lot for looking into this! The inference_diff tool is super for debugging the issue. Here is what I found so far:

- Following @srjoglekar246's lead, I tried the inference_diff and got slightly different errors on my Pixel 3XL. And I used tensorflow master branch and r2.2 branch and both produced different results. So my first question is: which branch I should use for Android deployment? I notice most users are using `org.tensorflow:tensorflow-lite-gpu:0.0.0-nightly` arr, but there is also `org.tensorflow:tensorflow-lite-gpu:2.2.0` arr. I tried both in my app, the results are very different on GPU delegation but same on CPU. 

- I also tried to allow GPU delegation to each single operation. Surprisingly, the first MUL operation already produced some error. I trimmed out all the operations but the first MUL operation and convert this tiny model to TFLite. Running on inference_diff and I got same error already, using  `gpu_precision_loss_allowed=true or false`. 

`output_errors {
      output_errors {
      max_value: 0.0021719106
      min_value: 0.00214503519
      avg_value: 0.0021564289927482605
      std_deviation: 1.0854229e-05
    }`

- Furthermore, I noticed that the inference_diff tool uses random inputs of [-1, 1] while my original input is [-127, 127]. If I hardcoded the inputs to all -127, I got even much bigger error with GPU delegation. Here is one line of code change in  [https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/lite/tools/evaluation/stages/inference_profiler_stage.cc#L51] 

`    data->push_back(static_cast<T>(-127.0f));
    //data->push_back(static_cast<T>(rand_float));`

Here is the error I got:

**Comment by ShiyongL** [comment]
``

**Comment by ShiyongL** [comment]
Sorry I was accidently closed this issue but reopened it again.

**Comment by srjoglekar246** [comment]
@ShiyongL for the AAR, it depends on whether you would prefer to use the TF nightly branch or the stabler TF 2.2 master branch for your environment. TF2.2 will probably be safer, but not contain any recent patches you might need. If the results are very different, can you explain how? Are the results with the 2.2 branch more correct? Then the issue might have been introduced recently.

About the 1-op model diff: In most delegates, there will be some diff v/s CPU, mainly because of the accumulators/optimizations used. For example, even with MobileNet v1 (where the GPU gives comparable top-k accuracy), there is a diff of ~2e-5 for the output tensor.

Do you have a way to check the task-oriented 'accuracy' of the model? (for example, how well your app recognises objects for an object detection model). That might help us see how bad the actual performance is (maybe with/without the gpu_precision_loss param)

**Comment by ShiyongL** [comment]
> @ShiyongL for the AAR, it depends on whether you would prefer to use the TF nightly branch or the stabler TF 2.2 master branch for your environment. TF2.2 will probably be safer, but not contain any recent patches you might need. If the results are very different, can you explain how? Are the results with the 2.2 branch more correct? Then the issue might have been introduced recently.
> 
> About the 1-op model diff: In most delegates, there will be some diff v/s CPU, mainly because of the accumulators/optimizations used. For example, even with MobileNet v1 (where the GPU gives comparable top-k accuracy), there is a diff of ~2e-5 for the output tensor.
> 
> Do you have a way to check the task-oriented 'accuracy' of the model? (for example, how well your app recognises objects for an object detection model). That might help us see how bad the actual performance is (maybe with/without the gpu_precision_loss param)

@srjoglekar246 for the 1-op model, TF2.2 and TF nightly branch(I assume it is the Master branch) produced slightly results:
TF nightly:
`OutputDiff[0]: avg_error=0.00215609, std_dev=1.08443e-05, gpu_precision_loss = false`
`OutputDiff[0]: avg_error=0.00215643, std_dev=1.08542e-05, gpu_precision_loss = true`
TF 2.2 (exactly same results for gpu_precision_loss = true or false ):
`output_errors { output_errors { max_value: 0.0021719106 min_value: 0.00214503519 avg_value: 0.0021564289927482605 std_deviation: 1.0854229e-05 }`
- The diff is 2e-3, which much larger than ~2e-5 by MobileNet v1. 
- When the input is 100x larger (change input from -1, 1 to -127, 127), the error is 100x larger. If you could reproduce the error using the single op model, it should help @impjdi for further debugging. 


To answer your second question, I am using the official ArcFace model from InsigntFace [https://github.com/deepinsight/insightface](url) to recognize faces. The input is (112x112x3) image, output is a 512 tensor. 
- With the first MUL operation, I was getting all NAN at my output. 
- After removing the MUL operations (which I do the MUL to input first then passing the result to CNN model), I was getting more reasonable results but still quite different (for a sample image):
Outputs from CPU: [0.7939907, 0.5848467, -0.94,       -0.99,           ..., 0.3934663, 0.93600059]
Outputs from GPU: [0.7978515, 0.3818359, -0.96875, -0.9086914, ..., 0.1318359, 0.77441406] (with gpu_precision_loss_allowed=false)  TF2.2 and TF nightly produced exactly same results here (CPU or GPU.)
I would be very happy if the error is ~e-2.

**Comment by srjoglekar246** [comment]
The output differences between TF2.2 & nightly seem too tiny to show any change in behavior between the two (~1e-7). So we can assume that nothing has been broken since TF2.2

About the MUL op:
1. The MUL op provided by @ShiyongL has an input tensor being multiplied by a scalar value `0.007812`. The avg diff w/ CPU does rise as the input sampling range is increased from {-1, 1} to {-100, 100} or {-1000, 1000} (proportionately). However, the std deviation of the diff values  remains ~5-e6, which indicates that the problem might just be accumulation errors that increase as the input values grow. I audited the OpenCL kernel on the GPU delegate, and it seems a straight-forward multiplication.

2. Even when I run inference_diff on the arcface model by rejecting all MUL ops (leading to delegation of only 5 ops somewhere in the middle of the graph), I still see an avg error of ~0.02. So MUL might not be the only issue.

A couple of clarifying questions from my side:

1. Is -127, 127 the expected input range for the model?

2. I am not sure what the 512-element output tensor represents, so looking at raw diffs might be misleading. Have you tried plugging in the model into an application that post-processes the output as needed? And does the precision with GPU fall a lot even after post-processing? A simple example I know is the SSD MobileNet, which has >1 raw diffs (CPU vs the DSP delegate) when we use inference_diff directly, but the COCO metrics remain good enough.

**Comment by ShiyongL** [comment]
I agree with @srjoglekar246 that MUL is not the only issue. Since it is the first operation, it is easier to trim out the rest and focusing on this particular operation. Like @srjoglekar246 said, multiplication should be a straightforward operation, the difference should not be that big between CPU and GPU. It is super interesting to see the reason. 

Re 1: -127, 127 was the expected input. Multiply by 0.007812 will normalize the input to -1, 1, which might be a better input for numerical stability reason. As I have shown the a previous comment, the difference between CPU and GPU is still quite large. 
Re2: the 512 output tensor is a 512 vector. ArcFace would inference two images and get 2 512 vectors. Then sure the distance between the two vectors. If the two vectors are similar, the two images are considered to be matched. I am not sure what you meant by >1 raw diffs for SSD MobileNet, which has multiple outputs like class, score and bboxes. I agree we should not compare just the raw values but the error ratio should be relatively low, like less than 1% due to floating point precision. 

I reported this issue when I got NAN as the output. Now we figured out the MUL is causing a big issue. After removing the MUL operation, I could get some more reasonable results, though the error is still not small and I am not totally convinced that this is caused by numerical stability. I will take the results anyhow and run the evaluation tomorrow. Will report back. 

Thanks a lot!

**Comment by ShiyongL** [comment]
@srjoglekar246 The errors between CPU and GPU are so large that it deteriorates the evaluation performance, when the inference results are used to determine whether one face image is similar to another. 
It would be greatly appreciated if you or other engineers could look into it!

**Comment by srjoglekar246** [comment]
@ShiyongL , I have asked the GPU folks internally to take a look at this model/issue. Will keep you posted on the details :-).

On that note, did you try any other delegates for this? (for example, XNNPack?)

**Comment by ShiyongL** [comment]
> @ShiyongL , I have asked the GPU folks internally to take a look at this model/issue. Will keep you posted on the details :-).
> 
> On that note, did you try any other delegates for this? (for example, XNNPack?)


Thanks a lot @srjoglekar246 ! 

I haven't tried XNNPack. Will check it out.

**Comment by srjoglekar246** [comment]
@ShiyongL we just fixed this on our end (will probably take a day or so to reflect in nightly). There were a couple of small issues with some GPU graph optimizations & model parsing which are now fixed.

The error for input in [-127, 127] is now ~0.02 with precision loss, and ~1e-6 with precision loss disabled (the precision loss issues are likely due to numerical instability w/ FP16). Could you do check the inference results again and close the issue if things work now?

**Comment by ShiyongL** [comment]
Perfect! 
I are using org.tensorflow:tensorflow-lite-gpu:0.0.0-nightly ARR. Do you know how long does it take for the fix to be available to the ARR, for example at [JCenter](https://bintray.com/google/tensorflow/tensorflow-lite)? 
On the other hand, do you mind sharing a link to a PR so that I could make sure to have the fixed code?  
Thank you very much!

**Comment by srjoglekar246** [comment]
@ShiyongL it should reflect within 24 hours or so. ([This](https://github.com/tensorflow/tensorflow/commit/225bdf60f3c4f51ab5568a53d31b0799369a1b89) is the commit after which the model should work fine).

**Comment by ShiyongL** [comment]
Managed to validate the fix. Thank you very much!
